### PR TITLE
feat(http): pass optional Request object to BrowserXhr.build

### LINKED ISSUE #11544: BrowserXhr build function is not aware of the request
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report => search github for a similar issue or PR before submitting
[X] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior** 
BrowserXhr build function does not know about the request it is being built for.

**Expected behavior**
Ideally the request object would be passed to build(...) so that decisions could be made based on requested resource.  It would simply be a matter of extending BrowserXhr and plugging in a custom build method implementation.

**What is the motivation / use case for changing the behavior?**
In some cases you might want to use an iframe's xhr to avoid CORS requests on sub-domains.  You may not want ALL requests to use it though, only ones to a specific API.  By having access to the request you could inspect the requested URL and make decisions based on that.

* **Angular version:** 2.0.0-rc.6

* **Browser:** all
 
* **Language:** all

### PR DESCRIPTION
**Pull request by DzmitryShylovich**
Fixes https://github.com/angular/angular/issues/11544

### CODE REVIEW

File: tools/public_api_guard/http/index.d.ts [hunk 1]
```diff
@@ -10,8 +10,7 @@ export declare class BaseResponseOptions extends ResponseOptions {
 
 /** @experimental */
 export declare class BrowserXhr {
-    constructor();
-    build(): any;
+    build(request?: Request): any;
```
Comment (DzmitryShylovich) [inline tools/public_api_guard/http/index.d.ts]: Why don't we explicitly specify the return type here?

Comment (laco0416) [inline tools/public_api_guard/http/index.d.ts]: +1 to `: XMLHttpRequest`

**Comment by DzmitryShylovich** [comment]
@vicb ping

**Comment by vicb** [comment]
1. Not sure to understand the purpose of this change ? `BrowserXhr` sounds like it should return an `XMLHttpRequest` ? (ie should the decision be taken before) ?

2. We can now introduce breaking change but they must be **clearly stated** in the commit message

**Comment by DzmitryShylovich** [comment]
> Not sure to understand the purpose of this change ?

motivation from the original issue: 
> Ideally the request object would be passed to build(...) so that decisions could be made based on requested resource. It would simply be a matter of extending BrowserXhr and plugging in a custom build method implementation.

> We can now introduce breaking change

this is why I made `request` param optional. It shouldn't break any code.

**Comment by vicb** [comment]
Closing from now as I'm not sure this is the right fix.
This PR will remain linked from the issue if we need to merge this.

**Pull request closed without merging**